### PR TITLE
[#316] Added the support for custom tasks

### PR DESCRIPTION
**Pull request by qzminski**
See https://github.com/andres-montanez/Magallanes/issues/316

### CODE REVIEW

**Comment by qzminski** [comment]
By the way this PR I think that an exception should be thrown if the task does not extend ```AbstractTask``` or otherwise the task would not be registered and executed – and no error thrown. We should notify the developer immediately that the registered task is not executable.

**Comment by Toflar** [comment]
I'd try to have some sort of interface so it's not even required to add anything at all anymore ;-)
Imagine a `Mage\Task\AutoloadInterface` with no methods at all (or we could even make an official one that includes `execute()`, `getName()` and `getDescription()` which would be a good idea anyway).

All you need to do is something like

```php
foreach (get_declared_classes() as $class) {
    if (is_a($class, AutoloadInterface::class, true)) {
        // load me
    }
}
```

**Comment by andres-montanez** [comment]
Hi, yes I was thinking of adding an interface in case the abstract class is too bulky.
Also agree with about throwing an error if the custom can't be loaded.

The problem with get_declared_classes() is that it will only return the used classes (or those requested), iterate over the custom_tasks collection will be still needed and a way to get them declared, like class_exists does, which pings the autoloader.

**Comment by Toflar** [comment]
Ah of course, you're right. Thinking about using Composer to provide that information but it would require to have installed the dependencies using `--optimize-autoload` to have an array of available classes. I never do that during development...

**Comment by qzminski** [comment]
Then how about providing the task-list classes in the ```custom_tasks``` section? See example:

```yml
custom_tasks:
    - 'FoobarLibrary\TaskList'
```

```php
namespace FoobarLibrary;

class TaskList extends SomeTaskListInterface
{
    public function getTasks()
    {
        return [
            'FoobarLibrary\Task1',
            'FoobarLibrary\Task2',
        ];
    }
}
```

**Comment by Toflar** [comment]
Imho it should be named `TaskProviderInterface` and thus the key should be `task_providers:` instead of `custom_tasks`. I think that would be cool.

**Comment by andres-montanez** [comment]
It sounds interesting and full of potential, BUT that won't be easy to read. You will end up with two configuration files, and one is a class. It will be easy to loose track of which tasks are being loaded and which doesn't.

Also let's not get out of the scope, the idea is an easy way to define custom tasks and save writing the class name each time.

We can split this and also add a feature for a Task Provider which could return tasks accordingly to the environment. For example if env is Production, return ['pre-deploy' => [tasks], 'post-deploy' => [tasks]], that seams a lot more useful. You will end up with a split config but at least it could be dynamic :)

**Comment by qzminski** [comment]
That's true but I think it does not much matter which task is loaded until it's actually used. Also the full list of tasks could be printed once #347 is implemented.

Indeed the idea of custom tasks we should leave as is and this new feature would go to the task providers.

As a task provider I would rather see a class that returns the list of available classes but does not decide about the stage they are executed in. This is too environment specific to be abstracted to some reusable portion of code. We've came up with this idea as we have an extra library (included via composer) for some extra deployment tasks and it would be easier to just provide ```custom/foobar``` task in the config instead of the full class name (which can change anytime contrary to the rather static task name/id).

**Comment by andres-montanez** [comment]
@qzminski I think we can meet both requirements, the provider can pre register the classes, but also could inject them if configured to do so. I know that I will benefit from the later ;)

**Comment by qzminski** [comment]
Oh well, sounds good :) Shall I provide a PR?

**Comment by Toflar** [comment]
I think we have to differentiate here. A task provider makes a lot of sense if you have an extra library that you reuse a lot. `composer require` and adding the task provider class to `.mage.yml` would be all there is to do then. That's pretty cool. Still registering them globally makes sense if you have local, project specific tasks.

File: tests/Resources/testhost-custom-task.yml [hunk 1]
```diff
@@ -14,7 +14,9 @@ magephp:
                 - composer/install
```
Comment (Toflar) [inline tests/Resources/testhost-custom-task.yml]: Sorry but this test is nonsense. It increases coverage but it does not test anything at all. You might as well just drop it. Please add a test that checks if `CustomTask` was really registered :)

**Comment by andres-montanez** [comment]
Closing the PR due to coverage drop.
Feature will be addressed in v3.2.
Regards.

**Pull request closed without merging**